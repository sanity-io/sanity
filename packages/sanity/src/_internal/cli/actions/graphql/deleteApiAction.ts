import type {SanityClient} from '@sanity/client'
import type {CliCommandArguments, CliCommandContext} from '@sanity/cli'
import yargs from 'yargs/yargs'
import {hideBin} from 'yargs/helpers'
import {getGraphQLAPIs} from './getGraphQLAPIs'

export interface DeleteGraphQLApiFlags {
  project?: string
  dataset?: string
  tag?: string
  force?: boolean
}

function parseCliFlags(args: {argv?: string[]}) {
  return yargs(hideBin(args.argv || process.argv).slice(2))
    .option('api', {type: 'string'})
    .option('project', {type: 'string'})
    .option('dataset', {type: 'string'})
    .option('tag', {type: 'string', default: 'default'})
    .option('force', {type: 'boolean'}).argv
}

export default async function deleteGraphQLApi(
  args: CliCommandArguments<DeleteGraphQLApiFlags>,
  context: CliCommandContext,
): Promise<void> {
  // Reparsing CLI flags for better control of binary flags
  const flags = await parseCliFlags(args)
  const {apiClient, output, prompt} = context

  // Use explicitly defined flags where possible
  let projectId = flags.project
  let dataset = flags.dataset
  let tag = flags.tag

  // If specifying --api, use it for the flags not provided
  if (flags.api) {
    const apiDefs = await getGraphQLAPIs(context)
    const apiDef = apiDefs.find((def) => def.id === flags.api)
    if (!apiDef) {
      throw new Error(`GraphQL API "${flags.api}" not found`)
    }

    if (projectId) {
      output.warn(`Both --api and --project specified, using --project ${projectId}`)
    } else {
      projectId = apiDef.projectId
    }

    if (dataset) {
      output.warn(`Both --api and --dataset specified, using --dataset ${dataset}`)
    } else {
      dataset = apiDef.dataset
    }

    if (tag && apiDef.tag) {
      output.warn(`Both --api and --tag specified, using --tag ${tag}`)
    } else {
      tag = apiDef.tag || 'default'
    }
  }

  // If neither --api nor --project/dataset is specified, use the CLI config for values
  let client: SanityClient
  if (!projectId || !dataset) {
    client = apiClient({
      requireUser: true,
      requireProject: true,
    }).config({apiVersion: '2023-08-01'})

    projectId = projectId || client.config().projectId
    dataset = dataset || client.config().dataset
  } else {
    client = apiClient({requireProject: false, requireUser: true}).config({projectId, dataset})
  }

  const confirmMessage =
    tag === 'default'
      ? `Are you absolutely sure you want to delete the current GraphQL API connected to the "${dataset}" dataset in project ${projectId}?`
      : `Are you absolutely sure you want to delete the GraphQL API connected to the "${dataset}" dataset in project ${projectId}, tagged "${tag}"?`

  const confirmedDelete =
    flags.force ||
    (await prompt.single({
      type: 'confirm',
      message: confirmMessage,
      default: false,
    }))

  if (!confirmedDelete) {
    return
  }

  if (projectId !== client.config().projectId) {
    client = client.clone().config({projectId})
  }

  try {
    await client.request({
      url: `/apis/graphql/${dataset}/${tag}`,
      method: 'DELETE',
    })
  } catch (err) {
    throw err
  }

  output.print('GraphQL API deleted')
}
