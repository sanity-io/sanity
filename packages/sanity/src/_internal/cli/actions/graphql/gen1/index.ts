import {ApiSpecification, GeneratedApiSpecification} from '../types'
import {generateTypeFilters} from './generateTypeFilters'
import {generateTypeQueries} from './generateTypeQueries'

export default (extracted: ApiSpecification): GeneratedApiSpecification => {
  const filters = generateTypeFilters(extracted.types)
  const queries = generateTypeQueries(extracted.types, filters)
  const types = [...extracted.types, ...filters]
  return {types, queries, interfaces: extracted.interfaces, generation: 'gen1'}
}
