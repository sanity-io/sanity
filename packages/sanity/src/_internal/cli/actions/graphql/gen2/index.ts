import type {ApiSpecification, GeneratedApiSpecification, InputObjectType} from '../types'
import {generateTypeFilters} from './generateTypeFilters'
import {generateTypeSortings} from './generateTypeSortings'
import {generateTypeQueries} from './generateTypeQueries'

export default (extracted: ApiSpecification): GeneratedApiSpecification => {
  const filters = generateTypeFilters(extracted.types)
  const sortings = generateTypeSortings(extracted.types)
  const queries = generateTypeQueries(
    extracted.types,
    sortings.filter((node): node is InputObjectType => node.kind === 'InputObject'),
  )
  const types = [...extracted.types, ...filters, ...sortings]

  return {types, queries, interfaces: extracted.interfaces, generation: 'gen2'}
}
