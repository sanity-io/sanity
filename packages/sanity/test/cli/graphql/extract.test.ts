import {orderBy} from 'lodash'
import {extractFromSanitySchema} from '../../../src/_internal/cli/actions/graphql/extractFromSanitySchema'
import type {ApiSpecification} from '../../../src/_internal/cli/actions/graphql/types'
import testStudioSchema from './fixtures/test-studio'

describe('GraphQL - Schema extraction', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    jest.resetModules()
  })

  afterEach(() => {
    jest.runAllTimers()
  })

  /**
   * @jest-environment jsdom
   */
  it('Should be able to extract schema', () => {
    const extracted = extractFromSanitySchema(testStudioSchema, {
      nonNullDocumentFields: false,
    })

    expect(sortExtracted(extracted)).toMatchSnapshot()
  })
})

function sortExtracted(schema: ApiSpecification) {
  const interfaces = orderBy(schema.interfaces, (iface) => iface.name).map((iface) => ({
    ...iface,
    fields: orderBy(iface.fields, (field) => field.fieldName),
  }))

  const types = orderBy(schema.types, (type) => type.name).map((type) => ({
    ...type,
    fields: orderBy((type as any).fields, (field) => field.fieldName),
  }))

  return {interfaces, types}
}
