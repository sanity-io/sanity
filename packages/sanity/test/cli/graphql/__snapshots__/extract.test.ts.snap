// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL - Schema extraction Should be able to extract schema 1`] = `
Object {
  "interfaces": Array [
    Object {
      "description": "A Sanity document",
      "fields": Array [
        Object {
          "description": "Date the document was created",
          "fieldName": "_createdAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": "Document ID",
          "fieldName": "_id",
          "isNullable": true,
          "type": "ID",
        },
        Object {
          "description": "Current document revision",
          "fieldName": "_rev",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Document type",
          "fieldName": "_type",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Date the document was last modified",
          "fieldName": "_updatedAt",
          "isNullable": true,
          "type": "Datetime",
        },
      ],
      "kind": "Interface",
      "name": "Document",
    },
  ],
  "types": Array [
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": "Date the document was created",
          "fieldName": "_createdAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": "Document ID",
          "fieldName": "_id",
          "isNullable": true,
          "type": "ID",
        },
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": "Current document revision",
          "fieldName": "_rev",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Document type",
          "fieldName": "_type",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Date the document was last modified",
          "fieldName": "_updatedAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "children": Object {
            "type": "String",
          },
          "description": undefined,
          "fieldName": "awards",
          "kind": "List",
        },
        Object {
          "description": undefined,
          "fieldName": "bestFriend",
          "isReference": true,
          "type": "Author",
        },
        Object {
          "fieldName": "image",
          "type": "Image",
        },
        Object {
          "children": Object {
            "type": "Block",
          },
          "description": undefined,
          "fieldName": "minimalBlockRaw",
          "isRawAlias": true,
          "kind": "List",
          "originalName": "minimalBlock",
          "type": "JSON",
        },
        Object {
          "description": undefined,
          "fieldName": "name",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "role",
          "type": "String",
        },
      ],
      "interfaces": Array [
        "Document",
      ],
      "kind": "Type",
      "name": "Author",
      "originalName": "author",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "children": Object {
            "type": "Span",
          },
          "description": undefined,
          "fieldName": "children",
          "kind": "List",
        },
        Object {
          "description": undefined,
          "fieldName": "list",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "style",
          "type": "String",
        },
      ],
      "kind": "Type",
      "name": "Block",
      "originalName": "block",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "code",
          "type": "Text",
        },
        Object {
          "description": undefined,
          "fieldName": "filename",
          "type": "String",
        },
        Object {
          "children": Object {
            "type": "Float",
          },
          "description": undefined,
          "fieldName": "highlightedLines",
          "kind": "List",
        },
        Object {
          "description": undefined,
          "fieldName": "language",
          "type": "String",
        },
      ],
      "kind": "Type",
      "name": "Code",
      "originalName": "code",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "alpha",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "hex",
          "type": "String",
        },
        Object {
          "fieldName": "hsl",
          "type": "HslaColor",
        },
        Object {
          "fieldName": "hsv",
          "type": "HsvaColor",
        },
        Object {
          "fieldName": "rgb",
          "type": "RgbaColor",
        },
      ],
      "kind": "Type",
      "name": "Color",
      "originalName": "color",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_dataset",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_projectId",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_ref",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_weak",
          "type": "Boolean",
        },
      ],
      "kind": "Type",
      "name": "CrossDatasetReference",
      "originalName": "crossDatasetReference",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": "Date the document was created",
          "fieldName": "_createdAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": "Document ID",
          "fieldName": "_id",
          "isNullable": true,
          "type": "ID",
        },
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": "Current document revision",
          "fieldName": "_rev",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Document type",
          "fieldName": "_type",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Date the document was last modified",
          "fieldName": "_updatedAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": undefined,
          "fieldName": "title",
          "type": "String",
        },
      ],
      "interfaces": Array [
        "Document",
      ],
      "kind": "Type",
      "name": "DocumentActionsTest",
      "originalName": "documentActionsTest",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "asset",
          "isReference": true,
          "type": "SanityFileAsset",
        },
      ],
      "kind": "Type",
      "name": "File",
      "originalName": "file",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "alt",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "lat",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "lng",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "Geopoint",
      "originalName": "geopoint",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "a",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "h",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "l",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "s",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "HslaColor",
      "originalName": "hslaColor",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "a",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "h",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "s",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "v",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "HsvaColor",
      "originalName": "hsvaColor",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "asset",
          "isReference": true,
          "type": "SanityImageAsset",
        },
        Object {
          "fieldName": "crop",
          "type": "SanityImageCrop",
        },
        Object {
          "fieldName": "hotspot",
          "type": "SanityImageHotspot",
        },
      ],
      "kind": "Type",
      "name": "Image",
      "originalName": "image",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "asset",
          "isReference": true,
          "type": "MuxVideoAsset",
        },
      ],
      "kind": "Type",
      "name": "MuxVideo",
      "originalName": "mux.video",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "assetId",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "filename",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "playbackId",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "status",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "thumbTime",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "MuxVideoAsset",
      "originalName": "mux.videoAsset",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo1",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo11",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo13",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo15",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo17",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo19",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo3",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo5",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo7",
          "kind": "List",
        },
        Object {
          "children": Object {
            "type": "ObjectWithNestedArray",
          },
          "description": undefined,
          "fieldName": "arrayNo9",
          "kind": "List",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo0",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo10",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo12",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo14",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo16",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo18",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo2",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo4",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo6",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "fieldNo8",
          "type": "String",
        },
      ],
      "kind": "Type",
      "name": "ObjectWithNestedArray",
      "originalName": "objectWithNestedArray",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "calculated",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "nominal",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "optimistic",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "pessimistic",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "PertEstimate",
      "originalName": "pertEstimate",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": "Date the document was created",
          "fieldName": "_createdAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": "Document ID",
          "fieldName": "_id",
          "isNullable": true,
          "type": "ID",
        },
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": "Current document revision",
          "fieldName": "_rev",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Document type",
          "fieldName": "_type",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Date the document was last modified",
          "fieldName": "_updatedAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "children": Object {
            "type": "String",
          },
          "description": undefined,
          "fieldName": "primitives",
          "kind": "List",
        },
        Object {
          "description": undefined,
          "fieldName": "title",
          "type": "String",
        },
      ],
      "interfaces": Array [
        "Document",
      ],
      "kind": "Type",
      "name": "Poppers",
      "originalName": "poppers",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "a",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "b",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "g",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "r",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "RgbaColor",
      "originalName": "rgbaColor",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": "The unique ID for the asset within the originating source so you can programatically find back to it",
          "fieldName": "id",
          "type": "String",
        },
        Object {
          "description": "A canonical name for the source this asset is originating from",
          "fieldName": "name",
          "type": "String",
        },
        Object {
          "description": "A URL to find more information about this asset in the originating source",
          "fieldName": "url",
          "type": "String",
        },
      ],
      "kind": "Type",
      "name": "SanityAssetSourceData",
      "originalName": "sanity.assetSourceData",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": "Date the document was created",
          "fieldName": "_createdAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": "Document ID",
          "fieldName": "_id",
          "isNullable": true,
          "type": "ID",
        },
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": "Current document revision",
          "fieldName": "_rev",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Document type",
          "fieldName": "_type",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Date the document was last modified",
          "fieldName": "_updatedAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": undefined,
          "fieldName": "altText",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "assetId",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "description",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "extension",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "label",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "mimeType",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "originalFilename",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "path",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "sha1hash",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "size",
          "type": "Float",
        },
        Object {
          "fieldName": "source",
          "type": "SanityAssetSourceData",
        },
        Object {
          "description": undefined,
          "fieldName": "title",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "url",
          "type": "String",
        },
      ],
      "interfaces": Array [
        "Document",
      ],
      "kind": "Type",
      "name": "SanityFileAsset",
      "originalName": "sanity.fileAsset",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": "Date the document was created",
          "fieldName": "_createdAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": "Document ID",
          "fieldName": "_id",
          "isNullable": true,
          "type": "ID",
        },
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": "Current document revision",
          "fieldName": "_rev",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Document type",
          "fieldName": "_type",
          "isNullable": true,
          "type": "String",
        },
        Object {
          "description": "Date the document was last modified",
          "fieldName": "_updatedAt",
          "isNullable": true,
          "type": "Datetime",
        },
        Object {
          "description": undefined,
          "fieldName": "altText",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "assetId",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "description",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "extension",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "label",
          "type": "String",
        },
        Object {
          "fieldName": "metadata",
          "type": "SanityImageMetadata",
        },
        Object {
          "description": undefined,
          "fieldName": "mimeType",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "originalFilename",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "path",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "sha1hash",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "size",
          "type": "Float",
        },
        Object {
          "fieldName": "source",
          "type": "SanityAssetSourceData",
        },
        Object {
          "description": undefined,
          "fieldName": "title",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "uploadId",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "url",
          "type": "String",
        },
      ],
      "interfaces": Array [
        "Document",
      ],
      "kind": "Type",
      "name": "SanityImageAsset",
      "originalName": "sanity.imageAsset",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "bottom",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "left",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "right",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "top",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "SanityImageCrop",
      "originalName": "sanity.imageCrop",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "aspectRatio",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "height",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "width",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "SanityImageDimensions",
      "originalName": "sanity.imageDimensions",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "height",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "width",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "x",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "y",
          "type": "Float",
        },
      ],
      "kind": "Type",
      "name": "SanityImageHotspot",
      "originalName": "sanity.imageHotspot",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "blurHash",
          "type": "String",
        },
        Object {
          "fieldName": "dimensions",
          "type": "SanityImageDimensions",
        },
        Object {
          "description": undefined,
          "fieldName": "hasAlpha",
          "type": "Boolean",
        },
        Object {
          "description": undefined,
          "fieldName": "isOpaque",
          "type": "Boolean",
        },
        Object {
          "fieldName": "location",
          "type": "Geopoint",
        },
        Object {
          "description": undefined,
          "fieldName": "lqip",
          "type": "String",
        },
        Object {
          "fieldName": "palette",
          "type": "SanityImagePalette",
        },
      ],
      "kind": "Type",
      "name": "SanityImageMetadata",
      "originalName": "sanity.imageMetadata",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "fieldName": "darkMuted",
          "type": "SanityImagePaletteSwatch",
        },
        Object {
          "fieldName": "darkVibrant",
          "type": "SanityImagePaletteSwatch",
        },
        Object {
          "fieldName": "dominant",
          "type": "SanityImagePaletteSwatch",
        },
        Object {
          "fieldName": "lightMuted",
          "type": "SanityImagePaletteSwatch",
        },
        Object {
          "fieldName": "lightVibrant",
          "type": "SanityImagePaletteSwatch",
        },
        Object {
          "fieldName": "muted",
          "type": "SanityImagePaletteSwatch",
        },
        Object {
          "fieldName": "vibrant",
          "type": "SanityImagePaletteSwatch",
        },
      ],
      "kind": "Type",
      "name": "SanityImagePalette",
      "originalName": "sanity.imagePalette",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "background",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "foreground",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "population",
          "type": "Float",
        },
        Object {
          "description": undefined,
          "fieldName": "title",
          "type": "String",
        },
      ],
      "kind": "Type",
      "name": "SanityImagePaletteSwatch",
      "originalName": "sanity.imagePaletteSwatch",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "current",
          "type": "String",
        },
      ],
      "kind": "Type",
      "name": "Slug",
      "originalName": "slug",
      "type": "Object",
    },
    Object {
      "description": undefined,
      "fields": Array [
        Object {
          "description": undefined,
          "fieldName": "_key",
          "type": "String",
        },
        Object {
          "description": undefined,
          "fieldName": "_type",
          "type": "String",
        },
        Object {
          "children": Object {
            "type": "String",
          },
          "description": undefined,
          "fieldName": "marks",
          "kind": "List",
        },
        Object {
          "description": undefined,
          "fieldName": "text",
          "type": "String",
        },
      ],
      "kind": "Type",
      "name": "Span",
      "originalName": "span",
      "type": "Object",
    },
  ],
}
`;
